ring/ring.go

```diff
@@ -0,0 +1,100 @@
+package ring
+
+import (
+	"sync/atomic"
+	"time"
+)
+
+// Consumer is the user-defined object responsible for receiving and processing
+// elements in batches when buffers are drained.
+type Consumer interface {
+	Push([]uint64)
+}
+
+// Buffer is a singular ("stripe") lossy ring buffer.
+type Buffer struct {
+	cons Consumer
+	data []uint64
+	size uint64
+	head uint64
+	busy uint64
+}
+
+func newBuffer(size uint64, consumer Consumer) *Buffer {
+	return &Buffer{
+		cons: consumer,
+		data: make([]uint64, size),
+		size: size,
+	}
+}
+
+func (b *Buffer) push(element uint64) bool {
+	// increment head to get the next available position in the buffer, and
+	// check if we need to drain
+	if head := atomic.AddUint64(&b.head, 1); head >= b.size {
+		// the buffer is full, so attempt to get exclusive access to the data
+		// so we can push the data to the Consumer
+		if atomic.CompareAndSwapUint64(&b.busy, 0, 1) {
+			// push buffer contents to the consumer
+			b.cons.Push(append(b.data[:0:0], b.data...))
+			// since the old elements are cleared out, place the new element
+			// at the front of the buffer
+			b.data[0] = element
+			// reset buffer head and unlock the busy mutex
+			atomic.StoreUint64(&b.head, 0)
+			atomic.StoreUint64(&b.busy, 0)
```

Comment:
These two atomics are not composable, so you might have a few empty slots in the beginning due to a race (e.g. context switch). The consumer would need to be tolerant of absent values.